toplevel/topdirs.ml

```diff
@@ -549,6 +549,22 @@ let () =
     )
     "Print the signature of the corresponding type constructor."
 
+let () =
+  reg_show_prim "show_constructor"
+    (fun env loc id lid ->
+       let desc = Env.lookup_constructor ~loc Env.Positive lid env in
+       if (Ctype.equal env true [desc.cstr_res] [Predef.type_exn]) then
+         raise Not_found;
```

Comment:
You don't need to special case exceptions here, exception constructors are constructors.

Reply:
Their type definition would look like `type exn = ..` which is not very useful. They could be printed as `type exn += Not_found`.

Reply:
But don't I need to exclude the handling of exceptions in this show_constructors show_prim as they are dealt with in the show _exceptions show_prim - maybe the naming is wrong and it should be "show_non_exception_constructors"

Reply:
to be fair there is now some common code that can be put into a little function:
```
let is_exception_constructor env type_expr = 
  Ctype.equal env true [type_expr] [Predef.type_exn]
```

and then used in both show_exception and show_constructor

Reply:
It makes perfect sense for `#show_exception` toreject non-exception constructors, but I agree that `#show_constructor` should also work on exception constructors, which are "just" a special case of extensible-variant constructors. (It's fine if the printing is not as good as with `#show_exception`.)

Reply:
I read more of the codebase, and I understand the issue now: the way `#show` is implemented is by calling all the `#show_...` actions in turn, and including in the output the result of each of those that did not raise `Not_found`. For this reason, it is better to raise `Not_found` here, or the output will be duplicated.

@wyn could you add a comment at the raising site that explains this? If we rework the implementation of the generic/agnostic `#show` someday we may be able to lift this limitation.

Reply:
(One possible approach would be to make `show_constructor` work *well* for exception (print them in a better way than generic extensible variants), by reusing the `show_exception` printing logic, and then simply rebind `show_exception` as an alias `show_constructor` instead of declaring it as a separately-registered show primitive. Then `show_exception` would work on non-exception types, and `show_constructor` would work on exceptions.)

Reply:
@gasche is that the same thing as having one show_constructor function and just having logic that switches on whether you have an exception or not?

```
let show_constructor bla bla =
  if (check is exception constructor) then
    print out exception stuff
  else
    print out constructor stuff
``` 
That is maybe more easy to understand at the cost of being a slightly longer function

Reply:
Yes, this structure is fine. Note that this discussion turns out to be about a choice between two behaviors:

1. either `show_constructor` fails on exceptions, and `show_exception` supports only exceptions
2. or `show_constructor` works on exceptions, and `show_exception` (now simply an alias of `show_constructor`) supports exceptions and also non-exception constructors

None of those options are fully satisfying, but they work well with the current "try everything and aggregate the result" approach of the `#show` implementation. I'm not completely sure which one to prefer. If you are also not sure, then keeping the code as you have written it currently (with a comment at the raise point to explain why it's important to fail on exception constructors) is also a perfectly good option.

Reply:
i think since this is my first change I'lll stick with the simpler option of adding a new show_constructor function with comments, thanks for the help.